### PR TITLE
Notification setting

### PR DESCRIPTION
**Pull request by KateMengjunXia**
Replacing https://github.com/foam-framework/foam2/pull/2765 (changed relationship cardinality)

nanopay side: https://github.com/nanoPayinc/NANOPAY/pull/8314

### CODE REVIEW

File: src/foam/nanos/auth/User.js [hunk 1]
```diff
@@ -655,12 +671,38 @@ foam.RELATIONSHIP({
   }
 });
 
+foam.RELATIONSHIP({
```
Comment (jlhughes) [inline src/foam/nanos/auth/User.js]: Move these Relationships to   foam/nanos/auth/Relationships.js

File: src/foam/nanos/notification/NotificationSetting.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        notification = (Notification) notification.fclone();
+        notification.setId(0L);
+        notification.setUserId(user.getId());
+        notification.setBroadcasted(false);
+        notification.setGroupId(null);
+        try {
+          notificationDAO.put_(x, notification);
+        } catch (Throwable t) {
+          Logger logger = (Logger) x.get("logger");
+          logger.error("Failed to send notification: " + t);
```
Comment (jlhughes) [inline src/foam/nanos/notification/NotificationSetting.js]:
```suggestion
          logger.error("Failed to send notification: " + t, t);
```
Add the trailing ,t so it generates the exception stack trace.

File: src/foam/nanos/auth/User.js (hunk 1 above)
Comment (KateMengjunXia) [reply src/foam/nanos/auth/User.js]: Done

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/auth/Relationships.js [hunk 3]
```diff
@@ -43,3 +43,25 @@ foam.RELATIONSHIP({
...
+  targetModel: 'foam.nanos.notification.NotificationSetting',
+  forwardName: 'notificationSettingsForBusinessUsers',
+  inverseName: 'businessUser',
+  sourceProperty: {
+    hidden: true
+  }
+});
+
+foam.RELATIONSHIP({
+  cardinality: '1:*',
```
Comment (TW80000) [inline src/foam/nanos/auth/Relationships.js:59]: Is a user going to have multiple `NotificationSetting`s?

File: src/foam/nanos/auth/User.js [hunk 4]
```diff
@@ -589,6 +594,17 @@ foam.CLASS({
...
         return this.label();
       }
+    },
+    {
+      name: 'doNotify',
+      javaCode: `
+        DAO notificationSettingDAO = (DAO) x.get("notificationSettingDAO");
+
+        List<NotificationSetting> settings = ((ArraySink) getNotificationSettings(x).select(new ArraySink())).getArray();
+        for( NotificationSetting setting : settings ) {
```
Comment (TW80000) [inline src/foam/nanos/auth/User.js:604]:
```suggestion
        for ( NotificationSetting setting : settings ) {
```

File: src/foam/nanos/auth/User.js [hunk 5]
```diff
@@ -657,10 +673,14 @@ foam.RELATIONSHIP({
 
 foam.CLASS({
   package: 'foam.nanos.auth',
-  name: 'UserUserJunctionGroupRefinement',
+  name: 'UserUserJunctionRefinement',
   refines: 'foam.nanos.auth.UserUserJunction',
 
   properties: [
+    {
+      class: 'Long',
```
Comment (TW80000) [inline src/foam/nanos/auth/User.js]: I don't think this makes sense. Junctions are identified by the composite key of the `sourceId` and `targetId`. I'd be surprised if this change didn't break a bunch of stuff.

File: src/foam/nanos/auth/User.js [hunk 6]
```diff
@@ -657,10 +673,14 @@ foam.RELATIONSHIP({
 
 foam.CLASS({
   package: 'foam.nanos.auth',
-  name: 'UserUserJunctionGroupRefinement',
+  name: 'UserUserJunctionRefinement',
```
Comment (TW80000) [inline src/foam/nanos/auth/User.js]: What's the reason for this change? Just naming?

File: src/foam/nanos/nanos.js [hunk 7]
```diff
@@ -167,6 +167,8 @@ FOAM_FILES([
   { name: 'foam/nanos/notification/email/Status' },
   { name: 'foam/nanos/notification/push/PushService' },
   { name: 'foam/nanos/notification/push/FirebasePushService' },
+  { name: "foam/nanos/notification/Notifiable" },
+  { name: "foam/nanos/notification/NotificationSetting" },
```
Comment (TW80000) [inline src/foam/nanos/nanos.js]: Do we need to list things here if they're listed in `files.js` as well? I thought we only needed to list them in one or the other.

File: src/foam/nanos/notification/EmailSetting.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
+foam.CLASS({
```
Comment (TW80000) [inline src/foam/nanos/notification/EmailSetting.js:8]: Missing license header.

File: src/foam/nanos/notification/EmailSetting.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+              if ( foam.util.SafetyUtil.isEmpty(notification.getEmailName()) ) {
+                message.setSubject(notification.getTemplate());
+                message.setBody(notification.getBody());
+                EmailsUtility.sendEmailFromTemplate(x, null, message, null, null);
+              } else {
+                EmailsUtility.sendEmailFromTemplate(x, user, message, notification.getEmailName(), notification.getEmailArgs());
+              }
+            } catch(Throwable t) {
+              Logger logger = (Logger) x.get("logger");
+              logger.error("Error sending notification email message: "+message+". Error: " + t);
```
Comment (TW80000) [inline src/foam/nanos/notification/EmailSetting.js]:
```suggestion
              logger.error("Error sending notification email message: " + message + ". Error: " + t);
```

**Review by zahnno — APPROVED** [review]

File: src/services [hunk 10]
```diff
@@ -422,7 +422,7 @@ p({"class":"foam.nanos.boot.NSpec", "name":"emailDocService",                  "
...
-p({"class":"foam.nanos.boot.NSpec", "name":"agentJunctionDAO",                 "lazy":false, "serve":true, "description":"DAO responsible for storing agent authentication permission information.","serviceScript": "return new foam.dao.EasyDAO.Builder(x).setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setJournalName(\"agentJunctions\").setOf(foam.nanos.auth.UserUserJunction.getOwnClassInfo()).build();","client":"{\"of\":\"foam.nanos.auth.UserUserJunction\", \"remoteListenerSupport\": false}"})
+p({"class":"foam.nanos.boot.NSpec", "name":"agentJunctionDAO",                 "lazy":false, "serve":true, "description":"DAO responsible for storing agent authentication permission information.","serviceScript": "return new foam.dao.EasyDAO.Builder(x).setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setJournalName(\"agentJunctions\").setSeqNo(true).setOf(foam.nanos.auth.UserUserJunction.getOwnClassInfo()).build();","client":"{\"of\":\"foam.nanos.auth.UserUserJunction\", \"remoteListenerSupport\": false}"})
```
Comment (TW80000) [inline src/services]: Please convert this to multi-line format.

File: src/services [hunk 11]
```diff
@@ -733,6 +733,23 @@ p({
   "serve": true,
   "serviceClass": "foam.nanos.session.SimpleSessionService"
 })
+
+p({
+  "class":"foam.nanos.boot.NSpec",
```
Comment (TW80000) [inline src/services]: Please add a space after the colon for key-value pairs.

File: src/services [hunk 12]
```diff
@@ -733,6 +733,23 @@ p({
...
+  "name":"notificationSettingDAO",
+  "serviceScript":"""
+    return new foam.dao.EasyDAO.Builder(x)
+      .setSeqNo(true)
+      .setPm(true)
+      .setJournalType(foam.dao.JournalType.SINGLE_JOURNAL)
+      .setJournalName("notificationSettings")
+      .setOf(foam.nanos.notification.NotificationSetting.getOwnClassInfo())
+      .setAuthorize(false)
```
Comment (TW80000) [inline src/services]: Is this safe? Probably not. Any user with access to this service will be able to delete/modify the settings of other users.

File: src/foam/nanos/notification/Notifiable.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+foam.INTERFACE({
```
Comment (TW80000) [inline src/foam/nanos/notification/Notifiable.js:8]: Missing license header.

File: src/foam/nanos/notification/NotificationSetting.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
+foam.CLASS({
```
Comment (TW80000) [inline src/foam/nanos/notification/NotificationSetting.js:8]: Missing license header.

File: src/foam/nanos/auth/Relationships.js (hunk 3 above)
Comment (TW80000) [reply src/foam/nanos/auth/Relationships.js:59]: I see how this works now.

File: src/foam/nanos/auth/User.js (hunk 6 above)
Comment (KateMengjunXia) [reply src/foam/nanos/auth/User.js]: Yes, just naming, since this refinement is not just about the group.

File: src/foam/nanos/auth/User.js (hunk 5 above)
Comment (KateMengjunXia) [reply src/foam/nanos/auth/User.js]: It seems like `id` on` UserUserJunction` is required for the relationship between `UserUserJunction` and `NotificationSetting`, let me know if I am wrong.

Comment (TW80000) [reply src/foam/nanos/auth/User.js]: You're right, but `id` is defined when the junction is created. We don't need to set it via a refinement.

See https://github.com/foam-framework/foam2/blob/master/src/foam/dao/Relationship.js#L290-L308

Comment (TW80000) [reply src/foam/nanos/auth/User.js]: And as you can see from that link, the `id` of a junction is a combination of the `sourceId` and `targetId`.

File: src/foam/nanos/auth/User.js (hunk 4 above)
Comment (jlhughes) [inline src/foam/nanos/auth/User.js:604]:
```suggestion
        for ( NotificationSetting setting : settings ) {
```
